driver/annotations.md

```diff
@@ -0,0 +1,171 @@
+file. This is done by calling the `On` rule constructor and chaining calls to the
+different selectors, rule constructors (a rule can hold other rules itself) and
+predicates to build your AST to UAST transformator.
+
+To see how rules are built it's important to understand that a rule is composed
+of:
+
+- One or more *predicates* that will select what nodes will match the rule.
+- One of more *actions* that will apply to the matching nodes.
+- Zero or more *selectors* that will optionally select more nodes starting
+  from the current ones and apply the rule/s in its argument list.
+- Zero or more additional *rules* contained inside to the *selectors*.
+
+The rule constructor is the function `On(predicates...) *Rule`. As you can see
+in the signature, it takes one of more predicates and returns a rule pointer. This 
+allows for chaining of different methods taking a rule pointer. A general form
+for a rule definition could this be like:
+
+```go
+// Simple example:
+var r  := On(somePredicate).SomeAction(actionParams)
+
+// A more elaborate example:
+var r2 := On(otherPredicate).SomeSelector(
+    On(thirdPredicate).SomeAction(actionParams), // embedded rule
+    On(fouthPredicate).SomeSelector( // embedded rule with its own childs
+        On(fithPredicate).SomeAction(actionParams),
+    )).SomeAction(actionPArams)
+```
+
+#### Predicates
+
+Predicates are any function that follow the signature: 
+
+```go
+func(n *uast.Node) bool
+```
+
+Where uast.Node are the Nodes that will be iterated during the annotation
+process. You could define your own predicate functions but the ann.go file
+has already defined the ones that you'll probably need most of the time:
+
+- `Any()`: always returns true. It's usually used to match the first node in 
+  the native AST and add rules from there.
+
+- `Not/And/Or(predicate)`: will apply the expected boolean operations to 
+  the predicate(s) used as argument.
+
+- `HasInternalType(type string)`: checks if the InternalType of the node matches
+  the string provided. What key is the InternalType of a native AST node is 
+  defined in the `parser.ObjectToNoder.InternalTypeKey` field.
+
+- `HasToken(name string)`: checks if the node has a token matching the given
+  string. What key of the node is a Token is defined in the 
+  `parser.ObjectToNoder.TokenKeys` map.
+
+- `HasChild(predicate)`: will check if any of the first level children of
+  the current node matches the given predicate.
+
+- `HasProperty(key, value string)`: checks if the AST node has some specific key
+  and the key as the given value.
+
+- `HasInternalRole(key string)`: this is actually a very handy shortcut to 
+  `HasProperty(uast.InternalRoleKey, value)` that will check if the node has
+  an InternalRole matching the specified key. InternalRoles are
+  added to children of any object keys holding them.
+
+#### Actions
+
+- `Roles(roles... uast.Role)`: Will add the given UAST role to the currently
```

Comment:
I'm talking about function definition, not its signature. The tense comment above also applies. So `adds` instead of `Will add`